common/lib/common-utils/src/promises.ts

```diff
@@ -12,12 +12,25 @@ export class Deferred<T> {
     private rej: ((reason?: any) => void) | undefined;
     private completed: boolean = false;
 
-    constructor() {
+    /**
+     * Create the Deferred, initializing the backing Promise
+     * If there may be a case where reject is called but no one awaits or .catch's the Promise,
+     * pass onRejection to avoid triggering an unhandledRejection.
+     * @param onRejection - (Optional) callback to pass to .catch on the Promise to ensure it isn't left floating.
+     */
+    constructor(onRejection?: (e: any) => void) {
```

Comment:
Do you think that having this optional argument will result in more case where people will provide callback?
I think stylistically I'd rather see deferred.promise.catch(callback), as it's a bit clearer intention over new Deferred(callback)

Reply:
can't this just be done directly on the promise `deferred.promise.catch()` rather that building into the deferred class?

Reply:
Yes there is a back story :)  I'm helping out Dylan Johnson on the Loop team who found that nearly all 403/404 errors coming through ODSP driver also result in an unhandled rejection.  We tracked it down to `EpochTrackerWithRedemption.fetchAndParseAsJSON` and that class's use of `Deferred` in its field `treesLatestDeferral`.

We believe that what's causing this is that the "JoinSessionSyncWait" event (the only code that does anything with `this.treesLatestDeferral.promise`) doesn't occur as apparently expected, so the promise never awaited or .catch'd.

That code needs to be fixed.  Dylan or I will put out a PR this week with a proposal, it may be as simple as adding a .catch after creating the Deferred (like this does - it could use this instead).  But this may indicate some bug or misunderstanding in that code that warrants a deeper fix as well.

Back to the question of whether _this_ PR will do anything to help - I'm not sure :)  But using Deferred thwarts static analysis like no-floating-promise lint rule, so I'm wondering if this parameter would help people to pause and think about their use of Deferred and whether they want this safety net.

Maybe there's nothing to do within the code here and we just need to keep this in mind when tracking down unhandledRejections.

Here's the [TS Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBANiA5jAvDUkRwKYDoGIDcAUMQQBQBEKNttlAlCcQPRvsedfeesswBVAHIBhAPIBZCQFEhAFRhyAEgEkAyjACCQgCIwASsJhjR0mGz4wA7gEswAExBXcAQ3v3pAN2xgoAGRtoH2wAJyoAVzAACxcHHHsQ7AArbGAoG3BKABoYcnJsTwZUAD54JHIAAwF9AC4YABIAbwLcRJcIcABfCoZekh4Bwa5SYDh2iBgdbAAzGEbiGEWYAAdwgCM4G2AVuoAFEJAAW0DsAB4wcMO10OKSJZWQm08XKGwYNscwOABPd+wIOrkAAedQuV1CRRQpU8IBs9juS2Wj2erz+bnAPz+SUB2DqsW+kOhsPhCyWGGgIXCaRAYSK83u9ygUUCuGWqBgYGwVhg+yOJzyiQgOUSSUJc1JDIZTJZgvZgoRksZzIgrWScuSCoZnSYpEV5n4mhgAFp0C9gFEYHYYPYZqFEvZ0OAKVSoDSJQy2NKVctcMAzVE8tgGFDGhQKlNpnbsPY6k1sN0mNrNZ1dUtBVhvOQwKDLtcQnSYF7VRAs0wYCn3SLUlB8niwAS5oXlaqkvkyymK+TYA60JzuRHyDr7C3q1RECAQPY1t9sIxmGwYAAhY2mqDmmAuaavELoNrpMCIFhVtJ2ZARqP2PjDn1+tcBgg6oA) I used to highlight the unhandledRejection in this case.

Reply:
I mean my first thought was to do `constructor(suppressUnhandledRejections?: true)` and just do an empty catch, but that's not a good idea (moves in the wrong direction in terms of writing correct code).

I think I'm probably leaning towards doing nothing and let the unhandledRejections speak for themselves, as difficult as they are to track down.  Interested in more feedback though now that I explained the motivation in more detail.